### PR TITLE
refactor: remove `State` class

### LINKED ISSUE #15638: Remove `State` class
`State` class is deprecated and should be removed. PTT and MTT should be refactored to use `MerkleStateRoot` rather than `State` (class) as the root node.

### PR DESCRIPTION
**Pull request by thenswan**
**Description**

This PR replaces usages of `com.swirlds.platform.state.State` with `com.swirlds.platform.state.MerkleStateRoot`, so that the former can be removed.  
It also adds initialization code required for `MerkleStateRoot` to work.  
For example, the following code initializes the platform state in the instance of `MerkleStateRoot` for further use:

```
     NoOpMerkleStateLifecycles.NO_OP_MERKLE_STATE_LIFECYCLES.initPlatformState(state);
```

Fixes #15638

### CODE REVIEW

File: platform-sdk/platform-apps/demos/CryptocurrencyDemo/src/main/java/com/swirlds/demo/crypto/CryptocurrencyDemoState.java [hunk 1]
```diff
@@ -380,76 +385,14 @@ public void init(
...
+        super.init(platform, trigger, previousSoftwareVersion);
 
+        this.platform = (SwirldsPlatform) platform;
         if (trigger == InitTrigger.GENESIS) {
             genesisInit();
         }
     }
 
-    @Override
-    public void serialize(final SerializableDataOutputStream out) throws IOException {
```
Comment (imalygin) [inline platform-sdk/platform-apps/demos/CryptocurrencyDemo/src/main/java/com/swirlds/demo/crypto/CryptocurrencyDemoState.java:391]: Not sure how important this particular state implementation is, but this serialization/deserialization logic cannot be just removed. However, this class may be no longer relevant, and it doesn't make sense spend too much time fixing it

File: platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java [hunk 2]
```diff
@@ -233,26 +237,6 @@ private void handleTransaction(@NonNull final ConsensusTransaction transaction)
         runningSum += delta;
     }
 
-    /**
-     * {@inheritDoc}
-     */
-    @Override
```
Comment (imalygin) [inline platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/com/swirlds/demo/addressbook/AddressBookTestingToolState.java:239]: Same here. We may want to do here something similar to this commit https://github.com/hashgraph/hedera-services/pull/16352/commits/a53a4a064b2d487465415a132e46fbf985e268cc

File: platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/ConsistencyTestingToolMain.java [hunk 3]
```diff
@@ -45,6 +53,31 @@ public class ConsistencyTestingToolMain implements SwirldMain {
...
+        }
+    }
+
+    private static void registerPlatformClasses(ConstructableRegistry constructableRegistry) {
+        logger.info(STARTUP.getMarker(), "Registering PlatformState classes with ConstructableRegistry");
+        final var schema = new V0540PlatformStateSchema();
+        final StateDefinition def = schema.statesToCreate().iterator().next();
+        final var md = new StateMetadata<>(PlatformStateService.NAME, schema, def);
+        registerWithSystem(md, constructableRegistry);
+    }
```
Comment (anthony-swirldslabs) [inline platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/ConsistencyTestingToolMain.java:79]: Is there a reason not to use the `lifecycles` implementation for this registration? In the AB->Roster refactoring fixes, I have to initialize the RosterState/RosterMap in addition to the PlatformState in each of such states. Having this logic duplicated in multiple places makes this hard, and also can lead to discrepancies/bugs. It would be best to consolidate this logic in a single place (such as the `lifecycles` class or similar).

This comment applies to similar code in other testing state classes in this fix.

File: platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/ConsistencyTestingToolState.java [hunk 4]
```diff
@@ -269,5 +279,8 @@ public void handleConsensusRound(final @NonNull Round round, final @NonNull Plat
         stateLong = NonCryptographicHashing.hash64(stateLong, round.getRoundNum());
 
         transactionHandlingHistory.processRound(ConsistencyTestingToolRound.fromRound(round, stateLong));
+
+        setChild(ROUND_HANDLED_INDEX, new StringLeaf(Long.toString(roundsHandled)));
+        setChild(STATE_LONG_INDEX, new StringLeaf(Long.toString(stateLong)));
```
Comment (anthony-swirldslabs) [inline platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/ConsistencyTestingToolState.java:284]: Is there a reason not to use the PlatformState here instead of introducing these two particular leaves?

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PayloadCfgSimple.java [hunk 5]
```diff
@@ -160,10 +167,21 @@ private static class ClassVersion {
...
+        //                StackTrace.getStackTrace());
+    }
 
     private PayloadCfgSimple(final PayloadCfgSimple sourcePayload) {
         super(sourcePayload);
+        //        logger.info(
+        //                LOGM_DEMO_INFO,
+        //                "PayloadCfgSimple.<init>(COPY), immutable: {}, StackTrace: {}",
+        //                isImmutable(),
+        //                StackTrace.getStackTrace());
```
Comment (anthony-swirldslabs) [inline platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PayloadCfgSimple.java:184]: Please remove the commented-out code (or please uncomment it if it's useful.)

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PayloadCfgSimple.java [hunk 6]
```diff
@@ -381,6 +399,11 @@ public void setRatioOfFCMTransaction(final float ratioOfFCMTransaction) {
      */
     @Override
     public PayloadCfgSimple copy() {
+        //        logger.info(
+        //                LOGM_DEMO_INFO,
+        //                "PayloadCfgSimple.copy(), immutable: {}, StackTrace: {}",
+        //                isImmutable(),
+        //                StackTrace.getStackTrace());
```
Comment (anthony-swirldslabs) [inline platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PayloadCfgSimple.java:406]: Please remove the commented-out code.

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolMain.java [hunk 7]
```diff
@@ -136,6 +146,39 @@ public class PlatformTestingToolMain implements SwirldMain {
...
+        }
+    }
+
+    private static void registerPlatformClasses(ConstructableRegistry constructableRegistry) {
+        logger.info(STARTUP.getMarker(), "Registering PlatformState classes with ConstructableRegistry");
+        final var schema = new V0540PlatformStateSchema();
+        final StateDefinition def = schema.statesToCreate().iterator().next();
+        final var md = new StateMetadata<>(PlatformStateService.NAME, schema, def);
+        registerWithSystem(md, constructableRegistry);
+    }
```
Comment (anthony-swirldslabs) [inline platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolMain.java:180]: Same as above - please consider implementing this in the `lifecycles` class instead of replicating this logic.

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolState.java [hunk 8]
```diff
@@ -243,8 +244,16 @@ public class PlatformTestingToolState extends PartialNaryMerkleInternal implemen
...
     private long transactionsIgnoredByExpectedMap = 0;
 
-    public PlatformTestingToolState() {
-        super(ChildIndices.CHILD_COUNT);
+    public PlatformTestingToolState(
+            @NonNull final MerkleStateLifecycles lifecycles,
+            @NonNull final Function<SemanticVersion, SoftwareVersion> versionFactory) {
+        super(lifecycles, versionFactory);
+
+        //        logger.info(
```
Comment (anthony-swirldslabs) [inline platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolState.java:252]: There's a lot of commented-out code in this class. Please either remove it, or uncomment it.

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolState.java [hunk 9]
```diff
@@ -650,7 +666,14 @@ public synchronized ExpectedFCMFamily getStateExpectedMap() {
...
     @Override
     public synchronized PlatformTestingToolState copy() {
+        //        logger.info(
+        //                LOGM_DEMO_INFO,
+        //                "PlatformTestingToolState.copy() immutable: {}, StackTrace: {}",
+        //                isImmutable(),
+        //                StackTrace.getStackTrace());
+
         throwIfImmutable();
+        //        setImmutable(true);
```
Comment (anthony-swirldslabs) [inline platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolState.java:676]: I'm not sure if it's safe not to make this state immutable upon copying.

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolState.java [hunk 10]
```diff
@@ -1305,6 +1328,7 @@ public void init(
             @NonNull final Platform platform,
             @NonNull final InitTrigger trigger,
             @Nullable final SoftwareVersion previousSoftwareVersion) {
+        //        super.init(platform, trigger, previousSoftwareVersion);
```
Comment (anthony-swirldslabs) [inline platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolState.java:1331]: Please either remove or uncomment.

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/TransactionCounter.java [hunk 11]
```diff
@@ -102,6 +104,12 @@ public TransactionCounter(int nodeId) {
     public TransactionCounter() {}
 
     private TransactionCounter(final TransactionCounter sourceTransactionCounter) {
+        //        logger.info(
```
Comment (anthony-swirldslabs) [inline platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/TransactionCounter.java:107]: Again, there's some commented out code in this class. Please take care of it.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/util/NoOpMerkleStateLifecycles.java [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+    public List<StateChanges.Builder> initPlatformState(@NonNull final State state) {
+        logger.info(MarkerManager.getMarker("DEMO_INFO"), "Init Platform State...");
+
+        if (!(state instanceof MerkleStateRoot merkleStateRoot)) {
+            throw new IllegalArgumentException("Can only be used with MerkleStateRoot instances");
+        }
+        final var schema = new V0540PlatformStateSchema();
+        schema.statesToCreate().stream()
+                .sorted(Comparator.comparing(StateDefinition::stateKey))
```
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/util/NoOpMerkleStateLifecycles.java:60]: I see that this code is very similar to what we're having in the `FakeMerkleStateLifecycles` in test fixtures.

I suggest to find a way to extract the common code into a single class instead of replicating it. This will ensure the consistent behavior between demo apps and tests, and wherever else we happen to use this functionality.

File: platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/ConsistencyTestingToolMain.java (hunk 3 above)
Comment (imalygin) [reply platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/ConsistencyTestingToolMain.java]: The reason was that `NO_OP_MERKLE_STATE_LIFECYCLE` didn't take care of all necessary class registrations that I extracted to `StateUtils#registerWithSystem`. However, you have a good point - they should not be separated. So, I add a call of `StateUtils#registerWithSystem` into `NoOpMerkleStateLifecycles#initPlatformState` and removed the logic duplication

File: platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/ConsistencyTestingToolState.java (hunk 4 above)
Comment (imalygin) [reply platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/ConsistencyTestingToolState.java:284]: If I understand the idea of this class correctly, these fields are directly correspond to anything what the platform state has. Previously it was stored separately, so I decided to follow the pattern.

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PayloadCfgSimple.java (hunk 6 above)
Comment (imalygin) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PayloadCfgSimple.java]: Done

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolMain.java (hunk 7 above)
Comment (imalygin) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolMain.java]: Removed

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/module-info.java [hunk 13]
```diff
@@ -23,6 +23,8 @@
     exports com.swirlds.demo.virtualmerkle.config to
             com.fasterxml.jackson.databind;
 
+    //    requires com.swirlds.state.impl;
```
Comment (andrewb1269hg) [inline platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/module-info.java:26]: Can we remove this commented out code?

**Review by andrewb1269hg — CHANGES_REQUESTED** [review]
Following files look good:
`hedera-node/hedera-app-spi/build.gradle.kts`
`platform-sdk/platform-apps/demos/CryptocurrencyDemo/src/main/java/module-info.java`
`platform-sdk/platform-apps/demos/HelloSwirldDemo/src/main/java/module-info.java`
`platform-sdk/platform-apps/demos/StatsDemo/src/main/java/module-info.java`
`platform-sdk/platform-apps/tests/AddressBookTestingTool/src/main/java/module-info.java`
`platform-sdk/platform-apps/tests/ConsistencyTestingTool/build.gradle.kts`
`platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/module-info.java`
`platform-sdk/platform-apps/tests/ISSTestingTool/src/main/java/module-info.java`
`platform-sdk/platform-apps/tests/MigrationTestingTool/src/main/java/module-info.java`
`platform-sdk/platform-apps/tests/PlatformTestingTool/build.gradle.kts`
`platform-sdk/platform-apps/tests/StatsSigningTestingTool/src/main/java/module-info.java`
`platform-sdk/platform-apps/tests/StressTestingTool/src/main/java/module-info.java`
`platform-sdk/swirlds-state-impl/src/main/java/module-info.java`

Following has requested change:
`platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/module-info.java`

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/module-info.java (hunk 13 above)
Comment (imalygin) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/module-info.java:26]: Sure, will do

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PayloadCfgSimple.java (hunk 5 above)
Comment (imalygin) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PayloadCfgSimple.java]: Removed

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PayloadCfgSimple.java (hunk 6 above)
Comment (imalygin) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PayloadCfgSimple.java]: Removed

**Review by andrewb1269hg — DISMISSED** [review]

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolState.java (hunk 8 above)
Comment (imalygin) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolState.java]: I cleaned up `PlatformTestingToolState`

**Review by andrewb1269hg — APPROVED** [review]

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolState.java (hunk 9 above)
Comment (imalygin) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolState.java]: Yes, it should be there. Uncommented

File: platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolState.java (hunk 10 above)
Comment (imalygin) [reply platform-sdk/platform-apps/tests/PlatformTestingTool/src/main/java/com/swirlds/demo/platform/PlatformTestingToolState.java]: Removed

**Review by andrewb1269hg — APPROVED** [review]

File: platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/ConsistencyTestingToolState.java (hunk 4 above)
Comment (anthony-swirldslabs) [reply platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/ConsistencyTestingToolState.java:284]: Okay. But my point is that the platform state encapsulates this data in a dedicated PlatfromState singleton rather than maintains it in separate tree leaves. So I'm just wondering why do we have to maintain these leaves explicitly here instead of simply adding the regular PlatformState to the state that this particular tool uses.

The reason I'm asking this is I suspect I'll have to add support for RosterService data into this demo state as well. And I see that unlike other demo states, this one doesn't follow the regular pattern of creating a PlatformState singleton and stuff. And it's unclear why it doesn't do so, and how am I supposed to add support for RosterService to it in the future...

Comment (imalygin) [reply platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/ConsistencyTestingToolState.java:284]: > But my point is that the platform state encapsulates this data in a dedicated PlatfromState singleton rather than maintains it in separate tree leaves.

This is true but these values are _not_  a part of the state in production (unless I'm missing something). I can _repurpose_ some fields of `com.hedera.hapi.platform.state.PlatformState` object to store these values, but that would look even more confusing. I don' think it's a good idea. 

An alternative to this implementation would be to have a service defined with a corresponding protobuf object. That would fit the existing pattern for services better, but, at the same time, it would be an overkill for this particular case. I don't want to pollute the protocol with test objects.

Comment (anthony-swirldslabs) [reply platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/ConsistencyTestingToolState.java:284]: Oh, I didn't realized these particular two values aren't a part of the actual production PlatformState object. Thanks for clarifying this! It would be great to have a comment added in code here to clarify this as well, so as to remove any potential confusion for future readers of this code.

But that's no longer an issue for me. Please feel free to resolve the conversation. Thanks!

**Review by andrewb1269hg — APPROVED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/util/NoOpMerkleStateLifecycles.java (hunk 12 above)
Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/util/NoOpMerkleStateLifecycles.java]: Yes, it's a good idea. So, I got rid of this class and used `FakeMerkleStateLifecycles`. It caused some trouble with the build, but thanks to @jjohannes it's resolved now.

**Review by andrewb1269hg — APPROVED** [review]

**Review by imalygin — APPROVED** [review]

**Review by andrewb1269hg — APPROVED** [review]

File: gradle/plugins/src/main/kotlin/com.hedera.gradle.protobuf.gradle.kts [hunk 14]
```diff
@@ -35,6 +35,8 @@ protobuf {
 
 configurations.configureEach {
     if (name.startsWith("protobufToolsLocator") || name.endsWith("ProtoPath")) {
+        @Suppress("UnstableApiUsage")
+        shouldResolveConsistentlyWith(configurations.getByName("mainRuntimeClasspath"))
```
Comment (mhess-swl) [inline gradle/plugins/src/main/kotlin/com.hedera.gradle.protobuf.gradle.kts:39]: Just curious, why is this necessary?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java [hunk 15]
```diff
@@ -16,7 +16,7 @@
 
 package com.swirlds.platform.state;
 
-import static com.swirlds.logging.legacy.LogMarker.EXCEPTION;
+import static com.swirlds.logging.legacy.LogMarker.*;
```
Comment (mhess-swl) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:19]: Explicit enumeration is preferred over star imports

File: gradle/plugins/src/main/kotlin/com.hedera.gradle.protobuf.gradle.kts (hunk 14 above)
Comment (imalygin) [reply gradle/plugins/src/main/kotlin/com.hedera.gradle.protobuf.gradle.kts:39]: This is how @jjohannes fixed an issue caused by a corner case of a "prod" code (`PlatformTestingTool` module) depending on a test module `swirlds-platform-core-test-fixture`. It doesn't work without this code and fails with a really obscure error:
```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':PlatformTestingTool:extractIncludeProto'.
> Could not resolve all dependencies for configuration ':PlatformTestingTool:compileProtoPath'.
   > Could not find org.junit.jupiter:junit-jupiter-params:.
     Required by:
         project :PlatformTestingTool > project :swirlds-platform-core > project :swirlds-state-impl

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Get more help at https://help.gradle.org/.

BUILD FAILED in 6s
```

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java (hunk 15 above)
Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:19]: I have a follow up PR that requires this once. Can I fix it there?

Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:19]: I mean this PR https://github.com/hashgraph/hedera-services/pull/16477

File: gradle/plugins/src/main/kotlin/com.hedera.gradle.protobuf.gradle.kts (hunk 14 above)
Comment (mhess-swl) [reply gradle/plugins/src/main/kotlin/com.hedera.gradle.protobuf.gradle.kts:39]: Ok, thanks for the context!

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java (hunk 15 above)
Comment (mhess-swl) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:19]: Yep sounds good, thanks!

**Review by mhess-swl — APPROVED** [review]
LGTM for services 👍
